File: problems/404_sum_of_left_leaves.py
from typing import Self


class TreeNode:
    def __init__(
        self, val: int = 0, left: Self | None = None, right: Self | None = None
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumOfLeftLeaves(self, root: TreeNode | None) -> int:
        if not root:
            return 0

        val = 0
        stack = [root]

        while stack:
            node = stack.pop()
            if node.left:
                stack.append(node.left)
                if node.left.left is None and node.left.right is None:
                    val += node.left.val

            if node.right:
                stack.append(node.right)

        return val

File: problems/404-sum-of-left-leaves.ts
export {}

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null

  constructor(
    val: number = 0,
    left?: TreeNode | null,
    right?: TreeNode | null
  ) {
    this.val = val
    this.left = left ?? null
    this.right = right ?? null
  }
}

function sumOfLeftLeaves(root: TreeNode | null): number {
  if (!root) {
    throw new Error("Invalid input.")
  }

  let sum = 0
  const stack = [root]

  while (stack.length) {
    const node = stack.pop()!
    if (node.left && !node.left.left && !node.left.right) {
      sum += node.left.val
    }
    if (node.left) {
      stack.push(node.left)
    }
    if (node.right) {
      stack.push(node.right)
    }
  }

  return sum
}
